### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by chingiz-mazhitov**

### CODE REVIEW

File: src/main/java/com/andersen/Ticket.java [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+	}
+
+	@Override
+	public int getId() {
+		return this.id;
+	}
+
+	@Override
+	public void setId(int id) {
+
```
Comment (Alexmansar) [inline src/main/java/com/andersen/Ticket.java]: why its empty?

File: src/main/java/com/andersen/Identifiable.java [hunk 2]
```diff
@@ -0,0 +1,12 @@
+package com.andersen;
+
+public interface Identifiable {
```
Comment (Alexmansar) [inline src/main/java/com/andersen/Identifiable.java]: in this case we can use abstract class as well (add getter and setter here, and the field as well)

File: src/main/java/com/andersen/Ticket.java [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+	private double baggageLimit;
+
+	@NullableWarning
+	private BigDecimal price = BigDecimal.valueOf(20.50);
+
+	public Ticket() {
+		CheckNullFields.checkNullFields(this);
+	}
+
+	public Ticket(String concertHall, int eventCode, long timeStamp) {
```
Comment (Alexmansar) [inline src/main/java/com/andersen/Ticket.java:30]: CheckNullFields.checkNullFields(this); 
mayve we need also add this here? what do you think?

File: src/main/java/com/andersen/Ticket.java [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+		return this.sector;
+	}
+
+	public void setSector(char sector) {
+		this.sector = sector;
+	}
+
+	public String getConcertHall() {
+		return concertHall;
+	}
```
Comment (Alexmansar) [inline src/main/java/com/andersen/Ticket.java]: in sime metods you use this, in some not.. pls use the same (or **with** or without)

File: src/main/java/com/andersen/Ticket.java
Comment (Alexmansar) [inline src/main/java/com/andersen/Ticket.java:1]: Please update getters and setters see.
image
In the task we can only work with a few fields using getters and setters, but now you have added the ability to work with all fields

File: src/main/java/com/andersen/users/User.java [hunk 5]
```diff
@@ -0,0 +1,9 @@
+package com.andersen.users;
+
+import com.andersen.Ticket;
+
+public abstract class User {
+
+	protected Ticket ticket;
```
Comment (Alexmansar) [inline src/main/java/com/andersen/users/User.java:8]: how you will add this ticket for your object?

**Review by Alexmansar — CHANGES_REQUESTED** [review]
@chingiz-mazhitov Hi, please check some comments) Overall good job, just needs a little improvements ))

File: src/main/java/com/andersen/users/User.java (hunk 5 above)
Comment (chingiz-mazhitov) [reply src/main/java/com/andersen/users/User.java:7]: sure, it's needed

File: src/main/java/com/andersen/Ticket.java
Comment (chingiz-mazhitov) [reply src/main/java/com/andersen/Ticket.java:1]: does it mean that only "time" and "stadium sector" can have setters and getter, and other fields only getters to get them?

Comment (Alexmansar) [reply src/main/java/com/andersen/Ticket.java:1]: yes))

File: src/main/java/com/andersen/users/User.java (hunk 5 above)
Comment (chingiz-mazhitov) [reply src/main/java/com/andersen/users/User.java:8]: I see I need to add constructor here

File: src/main/java/com/andersen/Identify.java [hunk 6]
```diff
@@ -0,0 +1,14 @@
+package com.andersen;
+
+public abstract class Identify {
+
+	protected int id;
+
+	protected abstract int getId();
+
+	protected abstract void setId(int id);
```
Comment (Alexmansar) [inline src/main/java/com/andersen/Identify.java]: you can use not abstarct method here. just simple get and set))

File: src/main/java/com/andersen/Ticket.java [hunk 7]
```diff
@@ -49,16 +48,18 @@ public Ticket(int id, String concertHall, int eventCode, long timeStamp, boolean
...
 	public int getId() {
-		return this.id;
+		return super.id;
 	}
 
 	@Override
 	public void setId(int id) {
-
+		super.id = id;
 	}
```
Comment (Alexmansar) [inline src/main/java/com/andersen/Ticket.java]: so this one's unnecessary.)

File: src/main/java/com/andersen/users/Admin.java [hunk 8]
```diff
@@ -1,20 +1,21 @@
...
 	public int getId() {
-		return this.id;
+		return super.id;
 	}
 
 	@Override
 	public void setId(int id) {
-		this.id = id;
+		super.id = id;
 	}
```
Comment (Alexmansar) [inline src/main/java/com/andersen/users/Admin.java]: the same here))

File: src/main/java/com/andersen/TicketService.java [hunk 9]
```diff
@@ -42,10 +42,10 @@ public static void main(String[] args) {
         ticket2.share(phone, email);
 
         // demonstrate dynamic polymorphism for Client and Admin subclasses
-        User client = new Client();
+        User client = new Client(ticket2);
         client.printRole();
 
-        User admin = new Admin();
+        User admin = new Admin(ticket2);
         admin.printRole();
```
Comment (Alexmansar) [inline src/main/java/com/andersen/TicketService.java]: 👍 
Good that you use abstraction, not realization))

File: src/main/java/com/andersen/Ticket.java [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+	public Ticket() {
+		CheckNullFields.checkNullFields(this);
+	}
+
+	public Ticket(String concertHall, int eventCode, long timeStamp) {
+		this.concertHall = concertHall;
+		this.eventCode = eventCode;
+		this.timeStamp = timeStamp;
+
+		CheckNullFields.checkNullFields(this);
```
Comment (Alexmansar) [inline src/main/java/com/andersen/Ticket.java]: move to the top in thos method - first validation, after that creation)

File: src/main/java/com/andersen/Ticket.java [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+			this.concertHall = concertHall.substring(0, 10);
+		}
+		this.eventCode = (eventCode >= 100 && eventCode <= 999) ? eventCode : 100;
+		this.timeStamp = timeStamp;
+		this.isPromo = isPromo;
+		this.sector = sector;
+		this.baggageLimit = baggageLimit;
+		this.price = isPromo ? (price.subtract(BigDecimal.valueOf(DISCOUNT))) : price;
+
+		CheckNullFields.checkNullFields(this);
```
Comment (Alexmansar) [inline src/main/java/com/andersen/Ticket.java]: the same

File: src/main/java/com/andersen/Ticket.java [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+
+	@Override
+	public int getId() {
+		return super.id;
+	}
+
+	@Override
+	public void setId(int id) {
+		super.id = id;
+	}
```
Comment (Alexmansar) [inline src/main/java/com/andersen/Ticket.java]: also can be removed)

File: src/main/java/com/andersen/users/Admin.java [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+
+	@Override
+	public int getId() {
+		return super.id;
+	}
+
+	@Override
+	public void setId(int id) {
+		super.id = id;
+	}
```
Comment (Alexmansar) [inline src/main/java/com/andersen/users/Admin.java]: the same

**Review by Alexmansar — CHANGES_REQUESTED** [review]
@chingiz-mazhitov please, check new comments)) its not critical, just small updates))

File: .gitignore [hunk 14]
```diff
@@ -3,6 +3,9 @@ out/
 !**/src/main/**/out/
 !**/src/test/**/out/
 .idea/
+pom.xml
```
Comment (Alexmansar) [inline .gitignore]: this one should in git)) here we have configs

**Review by Alexmansar — APPROVED** [review]
@chingiz-mazhitov good job)) could be merged))

File: .gitignore (hunk 14 above)
Comment (chingiz-mazhitov) [reply .gitignore]: added pom file to git

**Pull request merged**